File: library/src/main/java/com/pengrad/telegrambot/model/request/InlineKeyboardButton.java
package com.pengrad.telegrambot.model.request;

import java.io.Serializable;
import java.util.Objects;

/**
 * Stas Parshin
 * 06 May 2016
 */
public class InlineKeyboardButton implements Serializable {
    private final static long serialVersionUID = 0L;

    private String text;
    private String url;
    private LoginUrl login_url;
    private String callback_data;
    private String switch_inline_query;
    private String switch_inline_query_current_chat;
    private CallbackGame callback_game;
    private Boolean pay;

    private InlineKeyboardButton() {
    }

    //todo can use only one optional field, make different constructors or static methods

    public InlineKeyboardButton(String text) {
        this.text = text;
    }

    public InlineKeyboardButton url(String url) {
        this.url = url;
        return this;
    }

    public InlineKeyboardButton loginUrl(LoginUrl loginUrl) {
        login_url = loginUrl;
        return this;
    }

    public InlineKeyboardButton callbackData(String callbackData) {
        callback_data = callbackData;
        return this;
    }

    public InlineKeyboardButton switchInlineQuery(String switchInlineQuery) {
        switch_inline_query = switchInlineQuery;
        return this;
    }

    public InlineKeyboardButton switchInlineQueryCurrentChat(String switchInlineQueryCurrentChat) {
        switch_inline_query_current_chat = switchInlineQueryCurrentChat;
        return this;
    }

    public InlineKeyboardButton callbackGame(String callbackGame) {
        callback_game = new CallbackGame();
        return this;
    }

    public InlineKeyboardButton pay() {
        this.pay = true;
        return this;
    }

    public String text() {
        return text;
    }

    public String url() {
        return url;
    }

    public String callbackData() {
        return callback_data;
    }

    public String switchInlineQuery() {
        return switch_inline_query;
    }

    public String switchInlineQueryCurrentChat() {
        return switch_inline_query_current_chat;
    }

    public CallbackGame callbackGame() {
        return callback_game;
    }

    public boolean isPay() {
        return pay != null ? pay : false;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        InlineKeyboardButton that = (InlineKeyboardButton) o;
        return Objects.equals(text, that.text) &&
                Objects.equals(url, that.url) &&
                Objects.equals(login_url, that.login_url) &&
                Objects.equals(callback_data, that.callback_data) &&
                Objects.equals(switch_inline_query, that.switch_inline_query) &&
                Objects.equals(switch_inline_query_current_chat, that.switch_inline_query_current_chat) &&
                Objects.equals(callback_game, that.callback_game) &&
                Objects.equals(pay, that.pay);
    }

    @Override
    public int hashCode() {
        return Objects.hash(text, url, login_url, callback_data, switch_inline_query, switch_inline_query_current_chat, callback_game, pay);
    }

    @Override
    public String toString() {
        return "InlineKeyboardButton{" +
                "text='" + text + '\'' +
                ", url='" + url + '\'' +
                ", login_url=" + login_url +
                ", callback_data='" + callback_data + '\'' +
                ", switch_inline_query='" + switch_inline_query + '\'' +
                ", switch_inline_query_current_chat='" + switch_inline_query_current_chat + '\'' +
                ", callback_game=" + callback_game +
                ", pay=" + pay +
                '}';
    }
}

File: library/src/main/java/com/pengrad/telegrambot/model/request/InlineKeyboardButton.kt
package com.pengrad.telegrambot.model.request

import com.pengrad.telegrambot.model.WebAppInfo

/**
 * Stas Parshin
 * 06 May 2016
 */
data class InlineKeyboardButton @JvmOverloads constructor(
    @get:JvmName("text") var text: String? = null,
    @get:JvmName("url") var url: String? = null,
    @get:JvmName("loginUrl") var loginUrl: LoginUrl? = null,
    @get:JvmName("callbackData") var callbackData: String? = null,
    @get:JvmName("switchInlineQuery") var switchInlineQuery: String? = null,
    @get:JvmName("switchInlineQueryCurrentChat") var switchInlineQueryCurrentChat: String? = null,
    @get:JvmName("switchInlineQueryChosenChat") var switchInlineQueryChosenChat: SwitchInlineQueryChosenChat? = null,
    @get:JvmName("callbackGame") var callbackGame: CallbackGame? = null,
    var pay: Boolean? = null,
    @get:JvmName("webApp") var webApp: WebAppInfo? = null,
    @get:JvmName("copyText") var copyText: CopyTextButton? = null,
) {

    fun url(url: String): InlineKeyboardButton {
        this.url = url
        return this
    }

    fun loginUrl(loginUrl: LoginUrl): InlineKeyboardButton {
        this.loginUrl = loginUrl
        return this
    }

    fun callbackData(callbackData: String): InlineKeyboardButton {
        this.callbackData = callbackData
        return this
    }

    fun switchInlineQuery(switchInlineQuery: String): InlineKeyboardButton {
        this.switchInlineQuery = switchInlineQuery
        return this
    }

    fun switchInlineQueryCurrentChat(switchInlineQueryCurrentChat: String): InlineKeyboardButton {
        this.switchInlineQueryCurrentChat = switchInlineQueryCurrentChat
        return this
    }

    fun switchInlineQueryChosenChat(switchInlineQueryChosenChat: SwitchInlineQueryChosenChat): InlineKeyboardButton {
        this.switchInlineQueryChosenChat = switchInlineQueryChosenChat
        return this
    }

    @Suppress("UNUSED_PARAMETER")
    fun callbackGame(callbackGame: String): InlineKeyboardButton {
        this.callbackGame = CallbackGame
        return this
    }

    fun pay(): InlineKeyboardButton {
        this.pay = true
        return this
    }

    fun webApp(webApp: WebAppInfo): InlineKeyboardButton {
        this.webApp = webApp
        return this
    }

    fun copyText(text: String): InlineKeyboardButton {
        return copyText(CopyTextButton(text))
    }

    fun copyText(copyText: CopyTextButton): InlineKeyboardButton {
        this.copyText = copyText
        return this
    }

    fun isPay(): Boolean {
        return pay ?: false
    }
}
